File: leet_BFS_0111_min_depth.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func minDepth(root *TreeNode) int {
    if root == nil {
		return 0
	}

	D := []*TreeNode{root}
	res := 0
	for len(D) > 0 {
		res++
		len := len(D) // level length
		for len > 0 {
			node := D[0]
			D = D[1:]
			if node.Left == nil && node.Right == nil {
				return res
			}
			if node.Left != nil {
				D = append(D, node.Left)
			}
			if node.Right != nil {
				D = append(D, node.Right)
			}
			len--
		}
	}
	return res
}

File: leet_BFS_0111_min_depth.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */

var minDepth = function(root) {
    if (!root)
        return (0)

    let E = [ root ]
    let depth = 1
    while (E.length != 0) {
        let s = E.length
        let i = -1
        while (++i < s) {
            let node = E.shift()
            if (!node.left && !node.right) {
                return (depth)
            }
            if (node.left)
                E.push(node.left)
            if (node.right)
                E.push(node.right)
        }
        ++ depth
    }
    return (depth)
};
